File: src/nats/jsm.rs
//! JetStream management over `$JS.API.*` — raw subject request/reply so
//! the response payload flows through unchanged to the HTTP client.

use bytes::Bytes;
use serde_json::Value;

use super::admin::{discover, request_json, AdminError};
use super::Cluster;

/// Cluster-wide JetStream snapshot via `$SYS.REQ.SERVER.PING.JSZ`. With a
/// system-account connection this returns per-server JS state including
/// every account, stream, and consumer the server is aware of.
pub async fn jsm_overview(cluster: &Cluster) -> Result<Vec<Value>, AdminError> {
    let payload = serde_json::to_vec(&serde_json::json!({
        "accounts": true,
        "streams": true,
        "consumer": true,
        "config": true,
        "raft": true,
    }))?;
    discover(
        cluster,
        "$SYS.REQ.SERVER.PING.JSZ".to_string(),
        Bytes::from(payload),
        cluster.discovery_timeout(),
    )
    .await
}

pub async fn list_streams(cluster: &Cluster, offset: u64) -> Result<Value, AdminError> {
    let payload = serde_json::to_vec(&serde_json::json!({ "offset": offset }))?;
    request_json(
        cluster,
        "$JS.API.STREAM.LIST".to_string(),
        Bytes::from(payload),
    )
    .await
}

pub async fn stream_info(cluster: &Cluster, name: &str) -> Result<Value, AdminError> {
    request_json(cluster, format!("$JS.API.STREAM.INFO.{name}"), Bytes::new()).await
}

/// Update an existing stream's configuration. The payload must be a full
/// `StreamConfig` JSON (NATS rejects partial updates). Only certain fields
/// can change without recreate — see NATS docs.
pub async fn update_stream(cluster: &Cluster, name: &str, config: Value) -> Result<Value, AdminError> {
    let payload = serde_json::to_vec(&config)?;
    request_json(
        cluster,
        format!("$JS.API.STREAM.UPDATE.{name}"),
        Bytes::from(payload),
    )
    .await
}

pub async fn purge_stream(cluster: &Cluster, name: &str) -> Result<Value, AdminError> {
    request_json(
        cluster,
        format!("$JS.API.STREAM.PURGE.{name}"),
        Bytes::new(),
    )
    .await
}

pub async fn delete_stream(cluster: &Cluster, name: &str) -> Result<Value, AdminError> {
    request_json(
        cluster,
        format!("$JS.API.STREAM.DELETE.{name}"),
        Bytes::new(),
    )
    .await
}

pub async fn list_consumers(
    cluster: &Cluster,
    stream: &str,
    offset: u64,
) -> Result<Value, AdminError> {
    let payload = serde_json::to_vec(&serde_json::json!({ "offset": offset }))?;
    request_json(
        cluster,
        format!("$JS.API.CONSUMER.LIST.{stream}"),
        Bytes::from(payload),
    )
    .await
}

pub async fn consumer_info(
    cluster: &Cluster,
    stream: &str,
    consumer: &str,
) -> Result<Value, AdminError> {
    request_json(
        cluster,
        format!("$JS.API.CONSUMER.INFO.{stream}.{consumer}"),
        Bytes::new(),
    )
    .await
}

pub async fn delete_consumer(
    cluster: &Cluster,
    stream: &str,
    consumer: &str,
) -> Result<Value, AdminError> {
    request_json(
        cluster,
        format!("$JS.API.CONSUMER.DELETE.{stream}.{consumer}"),
        Bytes::new(),
    )
    .await
}

File: internal/natsx/jsm.go
package natsx

import (
	"encoding/json"
	"fmt"
)

// ListStreams paginates STREAM.LIST. Bound to the *connecting* account
// (so it fails with err_code 10039 when run on a $SYS connection that has
// no JetStream).
func (c *Cluster) ListStreams(offset uint64) (json.RawMessage, error) {
	payload, err := json.Marshal(map[string]uint64{"offset": offset})
	if err != nil {
		return nil, fmt.Errorf("encode list payload: %w", err)
	}
	return c.requestJSON("$JS.API.STREAM.LIST", payload)
}

// StreamInfo returns a single stream's info.
func (c *Cluster) StreamInfo(name string) (json.RawMessage, error) {
	return c.requestJSON("$JS.API.STREAM.INFO."+name, nil)
}

// UpdateStream PUTs a new StreamConfig. The payload must be a full config —
// NATS rejects partial updates. Some fields cannot change without recreate
// (storage, retention, etc.).
func (c *Cluster) UpdateStream(name string, config json.RawMessage) (json.RawMessage, error) {
	return c.requestJSON("$JS.API.STREAM.UPDATE."+name, config)
}

// PurgeStream drops every message from a stream (config and consumers stay).
func (c *Cluster) PurgeStream(name string) (json.RawMessage, error) {
	return c.requestJSON("$JS.API.STREAM.PURGE."+name, nil)
}

// DeleteStream removes the stream entirely.
func (c *Cluster) DeleteStream(name string) (json.RawMessage, error) {
	return c.requestJSON("$JS.API.STREAM.DELETE."+name, nil)
}

// ListConsumers paginates CONSUMER.LIST for a stream.
func (c *Cluster) ListConsumers(stream string, offset uint64) (json.RawMessage, error) {
	payload, err := json.Marshal(map[string]uint64{"offset": offset})
	if err != nil {
		return nil, fmt.Errorf("encode list payload: %w", err)
	}
	return c.requestJSON("$JS.API.CONSUMER.LIST."+stream, payload)
}

// ConsumerInfo returns one consumer's info.
func (c *Cluster) ConsumerInfo(stream, consumer string) (json.RawMessage, error) {
	return c.requestJSON(fmt.Sprintf("$JS.API.CONSUMER.INFO.%s.%s", stream, consumer), nil)
}

// DeleteConsumer removes a consumer.
func (c *Cluster) DeleteConsumer(stream, consumer string) (json.RawMessage, error) {
	return c.requestJSON(fmt.Sprintf("$JS.API.CONSUMER.DELETE.%s.%s", stream, consumer), nil)
}

// JSMOverview asks every server for a full JetStream report via
// $SYS.REQ.SERVER.PING.JSZ with {accounts, streams, consumer, config, raft}.
// With a system-account connection this yields cluster-wide visibility into
// every account, stream, and consumer the cluster knows about.
func (c *Cluster) JSMOverview() ([]json.RawMessage, error) {
	payload, err := json.Marshal(map[string]bool{
		"accounts": true,
		"streams":  true,
		"consumer": true,
		"config":   true,
		"raft":     true,
	})
	if err != nil {
		return nil, fmt.Errorf("encode overview payload: %w", err)
	}
	return c.discover("$SYS.REQ.SERVER.PING.JSZ", payload, c.discoveryTimeout)
}
